clar2wasm/src/words/sequences.rs

```diff
@@ -469,54 +469,55 @@ impl ComplexWord for Concat {
         let memory = generator.get_memory()?;
 
         // Create a new sequence to hold the result in the stack frame
-        let ty = generator
+        let ret_ty = generator
             .get_expr_type(expr)
             .ok_or_else(|| GeneratorError::TypeError("concat expression must be typed".to_owned()))?
             .clone();
-        let (offset, _) = generator.create_call_stack_local(builder, &ty, false, true);
-
-        builder.local_get(offset);
-
-        // Traverse the lhs, leaving it on the data stack (offset, size)
-        let lhs = args.get_expr(0)?;
-        // WORKAROUND: typechecker issue for lists
-        generator.set_expr_type(lhs, ty.clone())?;
-        generator.traverse_expr(builder, lhs)?;
-
-        // Save the length of the lhs
-        let lhs_length = generator.module.locals.add(ValType::I32);
-        builder.local_tee(lhs_length);
+        let (ret_off, _) = generator.create_call_stack_local(builder, &ret_ty, false, true);
 
-        // Copy the lhs to the new sequence
-        builder.memory_copy(memory, memory);
+        let lhs_expr = args.get_expr(0)?;
+        let rhs_expr = args.get_expr(1)?;
 
-        // Load the adjusted destination offset
-        builder
-            .local_get(offset)
-            .local_get(lhs_length)
-            .binop(BinaryOp::I32Add);
+        let lhs_len = generator.module.locals.add(ValType::I32);
+        let rhs_len = generator.module.locals.add(ValType::I32);
 
-        // Traverse the rhs, leaving it on the data stack (offset, size)
-        let rhs = args.get_expr(1)?;
         // WORKAROUND: typechecker issue for lists
-        generator.set_expr_type(rhs, ty.clone())?;
-        generator.traverse_expr(builder, rhs)?;
-
-        // Save the length of the rhs
-        let rhs_length = generator.module.locals.add(ValType::I32);
-        builder.local_tee(rhs_length);
-
-        // Copy the rhs to the new sequence
+        generator.set_expr_type(lhs_expr, ret_ty.clone())?;
+        generator.set_expr_type(rhs_expr, ret_ty)?;
```

Comment:
This is something that has always bothered me here, so maybe it's a good time to change this: the type of the lists arguments shouldn't be exactly the same as `ret_ty`. When we do this, we stupidly set the arguments max len to the same size as `ret_ty`. So if the traversal of an argument has to allocate memory, it will allocate the size of `ret_ty` instead of the real max size of the argument.

Reply:
Should I just remove those two lines then? It seems to me like these shouldn't even be here given what you're saying.

Reply:
Tests pass with the lines removes

Reply:
Normally, you were lucky if the tests pass. if I'm not wrong, we should still have an issue if we try to concatenate two lists with types which would be for example `(list 1 (optional Notype))` and `(list 1 (optional int))`.